File: src/HallowedSummonerEmote.ts
import SummonerEmote from './SummonerEmote';

export default class HallowedSummonerEmote {
  id: number;
  inventoryIcon: string;
  uri: string;
  categories: string[];
  filename: string;
  translations: ITranslations;

  constructor(props: IHallowedSummonerEmoteProps) {
    const { id, inventoryIcon, translations, uri, categories, filename } = props;
    this.id = id;
    this.inventoryIcon = inventoryIcon;
    this.uri = uri;
    this.categories = categories;
    this.filename = filename;
    this.translations = translations;
  }

  static fromSummonerEmote(emote: SummonerEmote): HallowedSummonerEmote {
    const hallowedSummonerEmote = new HallowedSummonerEmote({
      id: emote.id,
      inventoryIcon: emote.inventoryIcon,
      uri: emote.uri,
      categories: emote.categories,
      filename: emote.isIcon ? emote.filename : "",
      translations: {}
    });
    return hallowedSummonerEmote;
  }
}

interface IHallowedSummonerEmoteProps {
  id: number;
  inventoryIcon: string;
  uri: string;
  categories: string[];
  filename: string;
  translations: ITranslations
}

interface ITranslations { 
  [localeIdentifier: string]: { name: string, description: string};
}

File: Models/HallowedSummonerEmote.cs
using System.Collections.Immutable;
using System.Text.Json.Serialization;

namespace Models;

public class HallowedSummonerEmote 
{
    [JsonPropertyName("id")]
    public uint Id { get; set; }
    [JsonPropertyName("inventoryIcon")]
    public string InventoryIcon { get; set; } = "";
    [JsonPropertyName("uri")]
    public string Uri { get; set; } = "";
    [JsonPropertyName("categories")]
    public ImmutableList<string> Categories { get; set; } = ImmutableList<string>.Empty;
    [JsonPropertyName("filename")]
    public string FileName { get; set; } = "";
    [JsonPropertyName("translations")]
    public Dictionary<string, EmoteTranslation> Translations { get; init; } = new();

    public static HallowedSummonerEmote FromSummonerEmote(SummonerEmote emote)
    {
        return new HallowedSummonerEmote
        {
            Id = emote.Id,
            InventoryIcon = emote.InventoryIcon,
            Uri = emote.Uri,
            Categories = emote.Categories,
            FileName = emote.IsIcon ? emote.FileName : "",
        };
    }
}

public class EmoteTranslation
{
    [JsonPropertyName("name")]
    public string Name { get; set; } = "";
    [JsonPropertyName("description")]
    public string Description { get; set; } = "";
}
